src/storage-controller/src/instance.rs

```diff
@@ -231,38 +295,263 @@ where
+                        "scheduling multi-replica ingestion"
+                    );
+                    ingestion_state.active_replicas.insert(replica_id);
+                }
+            }
+        }
+
+        // Now send all collected commands
+        if send_commands {
+            for (replica_id, ingestion_ids) in commands_by_replica {
+                let commands: Vec<RunIngestionCommand> = ingestion_ids
+                    .into_iter()
+                    .map(|id| RunIngestionCommand {
+                        id,
+                        description: self
+                            .get_ingestion_description(&id)
+                            .expect("missing ingestion description")
+                            .clone(),
+                    })
+                    .collect();
+
+                if !commands.is_empty() {
+                    let replica = self.replicas.get_mut(&replica_id).expect("missing replica");
+                    replica.send(StorageCommand::RunIngestions(commands));
+                }
+            }
+        }
+    }
+
+    /// Returns the ingestion description for the given ingestion ID, if it
+    /// exists.
+    ///
+    /// This function searches through the command history to find the most
+    /// recent RunIngestionCommand for the specified ingestion ID and returns
+    /// its description.  Returns None if no ingestion with the given ID is
+    /// found.
+    pub fn get_ingestion_description(
+        &self,
+        id: &GlobalId,
+    ) -> Option<mz_storage_types::sources::IngestionDescription<CollectionMetadata>> {
+        if !self.active_ingestions.contains_key(id) {
+            return None;
+        }
+
+        self.history.iter().rev().find_map(|command| {
+            if let StorageCommand::RunIngestions(cmds) = command {
+                cmds.iter()
+                    .find(|cmd| &cmd.id == id)
+                    .map(|cmd| cmd.description.clone())
+            } else {
+                None
+            }
+        })
+    }
+
+    /// Updates internal state based on incoming compaction commands.
+    fn absorb_compactions(&mut self, cmds: Vec<(GlobalId, Antichain<T>)>) {
+        tracing::debug!(?self.active_ingestions, ?cmds, "allow_compaction");
+
+        for (id, frontier) in cmds.iter() {
+            if frontier.is_empty() {
+                self.active_ingestions.remove(id);
+                self.ingestion_exports.remove(id);
+                self.active_exports.remove(id);
+            }
+        }
+    }
+
+    /// Returns the replicas that are actively running the given object (ingestion or export).
+    fn active_replicas(&mut self, id: &GlobalId) -> Box<dyn Iterator<Item = &mut Replica<T>> + '_> {
+        if let Some(ingestion_id) = self.ingestion_exports.get(id) {
+            // Right now, only ingestions can have per-replica scheduling
+            // decisions.
+            if let Some(ingestion) = self.active_ingestions.get(ingestion_id) {
```

Comment:
nit: feels cleaner to make this a match instead of an if let with an early return to avoid falling through